src/qibo/models/encodings.py

```diff
@@ -1186,3 +1330,103 @@ def _intermediate_gate(
     )
 
     return gate, lex_order, initial_string, phase_index
+
+
+def _sort_data_sparse(data, nqubits):
+    from qibo.quantum_info.utils import (  # pylint: disable=import-outside-toplevel
+        hamming_weight,
+    )
+
+    bitstrings, _data = [], []
+    for bitstring, elem in data:
+        if isinstance(bitstring, int) or "int" in str(type(bitstring)):
+            bitstring = f"{bitstring:0{nqubits}b}"
+        bitstrings.append(bitstring)
+        _data.append(elem)
+    _data = list(zip(bitstrings, _data))
+    _data = sorted(_data, key=lambda x: hamming_weight(x[0]))
+
+    data_sorted, bitstrings_sorted = [], []
+    for string, elem in _data:
+        bitstrings_sorted.append(np.array(list(string)).astype(int))
+        data_sorted.append(elem)
+
+    return data_sorted, bitstrings_sorted
+
+
+def _get_gate_sparse(
+    distance, difference, touched_qubits, complex_data, controls, hw_0, hw_1, theta, phi
+):
+    if distance == 1:
+        qubit = np.where(difference == 1)[0][0]
+        if qubit not in touched_qubits:
+            touched_qubits.append(qubit)
+        gate = (
+            gates.U3(qubit, 2 * theta, 2 * phi, 0.0).controlled_by(*controls)
+            if complex_data
+            else gates.RY(qubit, 2 * theta).controlled_by(*controls)
+        )
+    elif distance == 2 and hw_0 == hw_1:
+        qubits = [np.where(difference == -1)[0][0], np.where(difference == 1)[0][0]]
+        for qubit in qubits:
+            if qubit not in touched_qubits:
+                touched_qubits.append(qubit)
+        gate = _get_gate(
+            [qubits[0]],
+            [qubits[1]],
+            controls,
+            theta,
+            phi,
+            complex_data,
+        )
+    else:
+        qubits = [np.where(difference == -1)[0], np.where(difference == 1)[0]]
+        for row in qubits:
+            for qubit in row:
+                if qubit not in touched_qubits:
+                    touched_qubits.append(qubit)
+        gate = gates.GeneralizedRBS(qubits[0], qubits[1], theta, -phi).controlled_by(
+            *controls
+        )
+
+    return gate, touched_qubits
```

Comment:
```suggestion
    return gate
```
In principle you wouldn't need to return ``touched_qubits`` as it is a list passed as input that is grown by the function. I would either copy the input list and modify and return the copy, or grow directly the input list and not return it for better clarity.